minio/select_object_reader.py

```diff
@@ -0,0 +1,284 @@
+        payload_length = byte_int(total_byte_length) - \
+            byte_int(headers_byte_length) - int(16)
+        payload = message[12+ byte_int(headers_byte_length) : 12+byte_int(headers_byte_length) + payload_length ]
+        message_crc = message[12+ byte_int(headers_byte_length) +payload_length:12+ byte_int(headers_byte_length) +payload_length + 4]
+
+        if not validate_crc(total_byte_length + headers_byte_length, prelude_crc):
+            raise CRCValidationError(
+                {"Checksum Mismatch, MessageCRC of " + \
+                str(calcuate_crc(total_byte_length + headers_byte_length)) + \
+                " does not equal expected CRC of "+ str(byte_int(prelude_crc))})
+
+        if not validate_crc(message[0:len(message)-4], message_crc):
+            raise CRCValidationError(
+                    {"Checksum Mismatch, MessageCRC of " + \
+                        str(calcuate_crc(message)) + \
+                        " does not equal expected CRC of "+ \
+                        str(byte_int(message_crc))})
+
+        header_map = self.extract_header(header, byte_int(headers_byte_length))
+
+        if header_map["message-type"] == EVENT :
+            # Parse message only when event-type is Records, 
+            # Progress, Stats. Break the loop if event type is End
+            # Do nothing if event type is Cont
+            if header_map["event-type"] == RECORDS or \
+                header_map["event-type"] == PROGRESS or \
+                header_map["event-type"] == STATS:
+                self.parse_message(header_map, payload, payload_length, rec)
+
+            if header_map["event-type"] == END:
+                flag = True
+        if header_map["message-type"] == ERROR:
+            self.parse_message(header_map, payload, payload_length, rec)
+            flag = True
+        return flag
+
+    
+
+    def read(self, num_bytes):
+        '''
+        extract each record from the response body ... and buffer it.
+        send only up to requested bytes such as message[:num_bytes]
+        rest is buffered and added to the next iteration.
+        '''
+        if len(self.remaining_bytes) == 0 :
+            res =  self.extract_message()
+            if len(res) == 0:
+                return ""            
+            else :
+                self.remaining_bytes = res
+
+        if num_bytes < len(self.remaining_bytes):
+            result = self.remaining_bytes[:num_bytes]
+            del self.remaining_bytes[:num_bytes] 
+            return result.decode("utf-8")
+        else :
+            left_in_buffer = self.remaining_bytes[:len(self.remaining_bytes)]
+            del self.remaining_bytes[:len(left_in_buffer)]
+            return left_in_buffer.decode("utf-8")   
+
+    def stream(self, num_bytes):
+        '''
+        streams the response 
+        '''
+        while True :
+            x = self.read(num_bytes)
+            if x == "":
```

Comment:
this should be more like `if x == b''` - check with python2 and python3 both.

Reply:
This code uses 
```def byte_int(data_bytes):
    '''
    Convert bytes to big-endian integer
    '''
    return int.from_bytes(data_bytes, byteorder='big')
 ``` 
which is a python3 feature. so not testing with python2.